File: mayonez-base/src/main/java/slavsquatsuperstar/mayonez/graphics/sprites/GLSprite.java
package slavsquatsuperstar.mayonez.graphics.sprites;

import org.joml.Vector4f;
import slavsquatsuperstar.math.Vec2;
import slavsquatsuperstar.mayonez.GameObject;
import slavsquatsuperstar.mayonez.annotations.EngineType;
import slavsquatsuperstar.mayonez.annotations.UsesEngine;
import slavsquatsuperstar.mayonez.io.GLTexture;
import slavsquatsuperstar.util.Color;

/**
 * A component that draws an image at a {@link GameObject}'s position using the GL engine.
 *
 * @author SlavSquatSuperstar
 */
@UsesEngine(EngineType.GL)
public final class GLSprite extends Sprite {

    private final GLTexture texture;
    private final Color color;

    /**
     * Create a new GLSprite that renders an entire texture.
     *
     * @param texture a texture
     */
    GLSprite(GLTexture texture) {
        this.texture = texture;
        color = new Color(new Vector4f(255f)); // Opaque white
    }

    /**
     * Create a new GLSprite that only renders a color.
     *
     * @param color a color
     */
    GLSprite(Color color) {
        this.color = color;
        texture = null;
    }

    // Getters and Setters

    public Color getColor() {
        return this.color;
    }

    @Override
    public GLTexture getTexture() {
        return texture;
    }

    public Vec2[] getTexCoords() {
        return (texture == null) ? GLTexture.DEFAULT_TEX_COORDS : texture.getTexCoords();
    }

    @Override
    public GLSprite copy() {
        return (texture == null) ? new GLSprite(color) : new GLSprite(texture);
    }

}

File: mayonez-base/src/main/kotlin/mayonez/graphics/sprites/GLSprite.kt
package mayonez.graphics.sprites

import mayonez.graphics.*
import mayonez.graphics.textures.*
import mayonez.math.*
import mayonez.math.shapes.*
import mayonez.renderer.gl.*


/**
 * Draws a [GLTexture] using the GL engine. This class should not be
 * directly instantiated. Instead, call [Sprites.createSprite]. See
 * [Sprite] for more information.
 *
 * @author SlavSquatSuperstar
 */
@UsesEngine(EngineType.GL)
internal class GLSprite private constructor(
    private var texture: GLTexture?, private var color: Color
) : Sprite(), GLQuad {

    /**
     * Create a new GLSprite that renders an entire texture.
     *
     * @param texture a texture
     */
    internal constructor(texture: GLTexture?) : this(texture, DEFAULT_COLOR)

    /**
     * Create a new GLSprite that only renders a color.
     *
     * @param color a color
     */
    internal constructor(color: Color?) : this(null, color ?: DEFAULT_COLOR)

    // Sprite Properties

    override val imageWidth: Int = texture?.width ?: 0

    override val imageHeight: Int = texture?.height ?: 0

    // Sprite Methods

    override fun getColor(): Color = color

    override fun setColor(color: Color?) {
        this.color = color ?: DEFAULT_COLOR
    }

    override fun getTexture(): GLTexture? = texture

    override fun setTexture(texture: Texture?) {
        this.texture = texture as? GLTexture
    }

    override fun copy(): GLSprite = GLSprite(texture, color)

    // Renderable Methods

    override fun getVertexPositions(): Array<out Vec2?>? {
        // Render sprite at object center and rotate according to object
        // Background sprite will not have scale but will use spriteXf instead
        val objXf = transform.combine(getSpriteTransform())
        return Rectangle(objXf.position, objXf.scale, objXf.rotation).vertices
    }

}